Add KeyUsage, ExtendedKeyUsage, CipherSuite & Protocol to SSL diagnos…

sindhusp opened the pull request:
Closes #63784

sindhusp commented:
Hi @tvernum , I signed the Contributor Agreement successfully, but that check is failing on my PR. Any thoughts on what I could do to verify the signing?

tvernum commented:
@sindhusp The issue seems to be that the email address you used in your git commit is not the same email address as you used to sign the CLA.
_and_ your commit doesn't link back to your GitHub account so we can't verify it based on your github userid.

Specifically your CLA has a gmail address, but your commit has a `zeta.tech` address.
And, it doesn't look like GitHub knows about your zeta.tech address either, so it doesn't know that this: https://github.com/sindhusp/elasticsearch/commit/3d8562345254269163e23b1f0397a9f1e6e23e06 is you - the author on that commit is just a name, but if you compare it to [one of my commits](https://github.com/tvernum/elasticsearch/commit/d2a22464f04f89eb6582db74c0565951d6f33a7d) GitHub knows that it's me, and links it to my GitHub account.

I think what you need to do is make sure your GitHub profile has both your  `zeta.tech` email address and your GMail address.

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -406,4 +413,38 @@ private static boolean checkIssuer(X509Certificate certificate, X509Certificate
+    private static void addSessionDescription(SSLSession session, StringBuilder message) {
+        String cipherSuite = Optional.ofNullable(session).map(SSLSession::getCipherSuite).orElse("<unknown cipherSuite>");
+        String protocol = Optional.ofNullable(session).map(SSLSession::getProtocol).orElse("<unknown protocol>");
+        message.append("; the session supports the cipher suite [")
```
For brevity we can omit articles from the diagnostic
```suggestion
        message.append("; the session supports cipher suite [")
```

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -406,4 +413,38 @@ private static boolean checkIssuer(X509Certificate certificate, X509Certificate
+        String protocol = Optional.ofNullable(session).map(SSLSession::getProtocol).orElse("<unknown protocol>");
+        message.append("; the session supports the cipher suite [")
+            .append(cipherSuite)
+            .append("] and ")
+            .append("the protocol [")
```
```suggestion
            .append("] and protocol [")
```

tvernum commented on libs/ssl-config/src/test/java/org/elasticsearch/common/ssl/SslDiagnosticsTests.java:
```diff
@@ -63,7 +63,8 @@ public void testDiagnosticMessageWhenServerProvidesAFullCertChainThatIsTrusted()
         assertThat(message, Matchers.equalTo("failed to establish trust with server at [192.168.1.1];" +
-            " the server provided a certificate with subject name [CN=cert1] and fingerprint [3bebe388a66362784afd6c51a9000961a4e10050];" +
+            " the server provided a certificate with subject name [CN=cert1] and fingerprint [3bebe388a66362784afd6c51a9000961a4e10050] and no keyUsage and no extendedKeyUsage;" +
+            " the session supports the cipher suite [<unknown cipherSuite>] and the protocol [<unknown protocol>];" +
```
We need some of the tests to have real values for all of these.
As it is we're only testing the "no value" cases, which is the least important case. 

That will require regenerating some of the certs (and updating the fingerprints) to have key usage properties.
(see libs/ssl-config/src/test/resources/certs/README.txt)
and/or changing `mockCertificateWithIssuer` and `session` to provide mock values.

sindhusp replied:
I've edited the `mockCertificateWithIssuer` and `session` methods to provide mock values. I attempted to regenerate the cert yesterday without much success, so went with the mocking. Let me know if this works.

tvernum reviewed (commented):
Thanks for your contribution, I have some suggested changes.

sindhusp committed:
Add KeyUsage, ExtendedKeyUsage, CipherSuite & Protocol to SSL diagnostics

sindhusp commented:
@tvernum thanks, my CLA problem is sorted, I've amended my commit to point to my gmail ID. 
Thanks for the quick review, I'm addressing your comments.

sindhusp committed:
Review comments

* Fix formatting
* Add descriptive keyUsage, extendedKeyUsage msgs
* Mock cert's keyUsage, extendedKeyUsage
* Mock session's cipherSuite, protocol

sindhusp commented:
@tvernum, I have addressed your comments. Please let me know if you have more suggestions.

sindhusp commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -415,23 +440,40 @@ private static boolean isSelfIssued(X509Certificate certificate) {
+            return "no keyUsage";
+        }
+
+        final String[] keyUsageGlossary = {"digitalSignature", "nonRepudiation", "keyEncipherment",
```
Should I use an enum here too like I've done with `ExtendedKeyUsage`, for uniformity? I could define the index as the Enum field.

sindhusp commented:
A little nudge for review @tvernum :)

sindhusp commented:
@tvernum bump for review.. Is it okay to remind this way? Is there someone else I can loop in for this PR?

tvernum commented:
He @sindhusp, Pinging here is fine but we're just a bit short of time at the moment and this PR has had to take a back seat. I hope to have time next week.

tvernum commented:
Hi @sindhusp Sorry for taking so long on this - can you let me know if you're still interested in working on this?

If you would don't have time, then someone from Elastic can finish it.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -406,4 +438,58 @@ private static boolean checkIssuer(X509Certificate certificate, X509Certificate
+        List<String> keyUsageDescription = new ArrayList<>();
+        IntStream.range(0, keyUsage.length).forEach(i -> {
+            if (keyUsage[i]) {
+                keyUsageDescription.add(keyUsageGlossary[i]);
```
For safety we should check that `i < keyUsageGlossary.length`

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -406,4 +438,58 @@ private static boolean checkIssuer(X509Certificate certificate, X509Certificate
+            "dataEncipherment", "keyAgreement", "keyCertSign", "cRLSign", "encipherOnly",
+            "decipherOnly"};
+
+        List<String> keyUsageDescription = new ArrayList<>();
+        IntStream.range(0, keyUsage.length).forEach(i -> {
+            if (keyUsage[i]) {
+                keyUsageDescription.add(keyUsageGlossary[i]);
+            }
+        });
+        return keyUsageDescription.stream()
+            .reduce((a, b) -> a + ", " + b)
```
The style use have used here is fine, but I would have written it like this:
```suggestion
        String keyUsageDescription = IntStream.range(0, keyUsage.length)
          .filter(i -> keyUsage[i])
          .map(i -> i < keyUsageGlossary.length ? keyUsageGlossary[i] : "#" + i)
          .collect(Collectors.joining(", "));
```
And then tested whether that was empty or not.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -406,4 +438,58 @@ private static boolean checkIssuer(X509Certificate certificate, X509Certificate
+        try {
+            return Optional.ofNullable(certificate.getExtendedKeyUsage())
+                .map(keyUsage -> generateExtendedKeyUsageDescription(keyUsage))
+                .orElse("no extendedKeyUsage");
```
You have this `orElse` case both here, and in  `generateExtendedKeyUsageDescription` and I don't think that's necessary.

tvernum commented on libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java:
```diff
@@ -406,4 +438,58 @@ private static boolean checkIssuer(X509Certificate certificate, X509Certificate
+        String protocol = Optional.ofNullable(session)
+            .map(SSLSession::getProtocol)
+            .orElse("<unknown protocol>");
+        message.append("; the session supports cipher suite [")
```
```suggestion
        message.append("; the session uses cipher suite [")
```
Strictly, by the time we get to a session being established it's not really "supports", it's "has selected for use"

sindhusp commented:
@tvernum thanks for the comments, I didn't see it earlier. I shall address them today.

tvernum committed:
Merge branch 'master' into sslDiagnostics

tvernum committed:
Tweak implementation and add more tests

tvernum commented:
@elasticmachine test this please 🙏

tvernum committed:
Fix broken test

tvernum commented:
@elasticmachine test this please 🙏

tvernum commented:
@jkakavas I made some tweaks to this PR - can you give it a quick look?

jkakavas reviewed (approved):
LGTM, double checked key usage and extended key usage in relevant RFCs and messaging looks fine to me
